File: youtube-cli/fatorial.js
/**
 * Função que calcula o fatorial de um número inteiro não negativo
 * @param {number} n - O número para calcular o fatorial
 * @returns {number} - O fatorial de n
 * @throws {Error} - Se n não for um inteiro não negativo
 */
function calcularFatorial(n) {
    // Verifica se o número é inteiro não negativo
    if (n < 0 || !Number.isInteger(n)) {
        throw new Error("O número deve ser um inteiro não negativo");
    }
    
    // Caso base: 0! = 1 e 1! = 1
    if (n === 0 || n === 1) {
        return 1;
    }
    
    // Calcula o fatorial iterativamente
    let resultado = 1;
    for (let i = 2; i <= n; i++) {
        resultado *= i;
    }
    
    return resultado;
}

// Exemplo de uso:
console.log(calcularFatorial(5)); // Retorna 120
console.log(calcularFatorial(0)); // Retorna 1
console.log(calcularFatorial(1)); // Retorna 1
console.log(calcularFatorial(10)); // Retorna 3628800

// Exporta a função para poder ser utilizada em outros arquivos
if (typeof module !== 'undefined' && module.exports) {
    module.exports = calcularFatorial;
}

File: youtube-cli/fatorial.py
def fatorial(n):
    """
    Calcula o fatorial de um número inteiro não negativo.
    
    Args:
        n (int): Número inteiro não negativo
        
    Returns:
        int: O fatorial de n
        
    Raises:
        ValueError: Se n for negativo
        TypeError: Se n não for inteiro
    """
    if not isinstance(n, int):
        raise TypeError("O argumento deve ser um número inteiro")
    
    if n < 0:
        raise ValueError("O fatorial não está definido para números negativos")
    
    if n == 0 or n == 1:
        return 1
    
    resultado = 1
    for i in range(2, n + 1):
        resultado *= i
    
    return resultado

# Exemplo de uso:
# print(fatorial(5))  # Saída: 120
# print(fatorial(0))  # Saída: 1
# print(fatorial(1))  # Saída: 1